File: java/revenj-core/src/main/java/org/revenj/patterns/DataContext.java
package org.revenj.patterns;

import rx.Observable;

import java.io.IOException;
import java.util.*;

public interface DataContext {
	<T extends Identifiable> List<T> find(Class<T> manifest, Collection<String> uris);

	default <T extends Identifiable> Optional<T> find(Class<T> manifest, String uri) {
		List<T> found = find(manifest, Collections.singletonList(uri));
		return found.size() == 1 ? Optional.of(found.get(0)) : Optional.<T>empty();
	}

	default <T extends Identifiable> List<T> find(Class<T> manifest, String[] uris) {
		return find(manifest, Arrays.asList(uris));
	}

	<T extends DataSource> Query<T> query(Class<T> manifest, Specification<T> filter);

	default <T extends DataSource> Query<T> query(Class<T> manifest) {
		return query(manifest, null);
	}

	<T extends DataSource> List<T> search(Class<T> manifest, Specification<T> filter, Integer limit, Integer offset);

	default <T extends DataSource> List<T> search(Class<T> manifest) {
		return search(manifest, null, null, null);
	}

	default <T extends DataSource> List<T> search(Class<T> manifest, Specification<T> filter) {
		return search(manifest, filter, null, null);
	}

	default <T extends DataSource> List<T> search(Class<T> manifest, Specification<T> filter, int limit) {
		return search(manifest, filter, limit, null);
	}

	<T extends DataSource> long count(Class<T> manifest, Specification<T> filter);

	default <T extends DataSource> long count(Class<T> manifest) {
		return count(manifest, null);
	}

	<T extends DataSource> boolean exists(Class<T> manifest, Specification<T> filter);

	default <T extends DataSource> boolean exists(Class<T> manifest) {
		return exists(manifest, null);
	}

	<T extends AggregateRoot> void create(Collection<T> aggregates) throws IOException;

	default <T extends AggregateRoot> void create(T aggregate) throws IOException {
		create(Collections.singletonList(aggregate));
	}

	<T extends AggregateRoot> void updatePairs(Collection<Map.Entry<T, T>> pairs) throws IOException;

	default <T extends AggregateRoot> void update(T oldAggregate, T newAggregate) throws IOException {
		updatePairs(Collections.singletonList(new HashMap.SimpleEntry<>(oldAggregate, newAggregate)));
	}

	default <T extends AggregateRoot> void update(T aggregate) throws IOException {
		updatePairs(Collections.singletonList(new HashMap.SimpleEntry<>(null, aggregate)));
	}

	default <T extends AggregateRoot> void update(Collection<T> aggregates) throws IOException {
		Collection<Map.Entry<T, T>> collection = new ArrayList<>(aggregates.size());
		for (T item : aggregates) {
			collection.add(new AbstractMap.SimpleEntry<>(null, item));
		}
		updatePairs(collection);
	}

	<T extends AggregateRoot> void delete(Collection<T> aggregates) throws IOException;

	default <T extends AggregateRoot> void delete(T aggregate) throws IOException {
		delete(Collections.singletonList(aggregate));
	}

	<T extends DomainEvent> void submit(Collection<T> events);

	default <T extends DomainEvent> void submit(T event) {
		submit(Collections.singletonList(event));
	}

	<T extends DomainEvent> void queue(Collection<T> events);

	default <T extends DomainEvent> void queue(T event) {
		queue(Collections.singletonList(event));
	}

	<T> T populate(Report<T> report);

	<T extends Identifiable> Observable<DataChangeNotification.TrackInfo<T>> track(Class<T> manifest);

	<T extends ObjectHistory> List<History<T>> history(Class<T> manifest, Collection<String> uris);

	default <T extends ObjectHistory> Optional<History<T>> history(Class<T> manifest, String uri) {
		List<History<T>> found = history(manifest, Collections.singletonList(uri));
		return found.size() == 1 ? Optional.of(found.get(0)) : Optional.<History<T>>empty();
	}

	default <T extends ObjectHistory> List<History<T>> history(Class<T> manifest, String[] uris) {
		return history(manifest, Arrays.asList(uris));
	}
}

File: scala/revenj-core/src/main/scala/net/revenj/patterns/DataContext.scala
package net.revenj.patterns

import scala.concurrent.Future
import scala.reflect.runtime.universe.TypeTag

trait DataContext {
  def find[T <: Identifiable : TypeTag](uris: Seq[String]): Future[IndexedSeq[T]]

  def find[T <: Identifiable : TypeTag](uri: String): Future[Option[T]] = {
    import scala.concurrent.ExecutionContext.Implicits.global
    find[T](Seq(uri)).map(_.headOption)
  }

  def search[T <: DataSource : TypeTag](filter: Option[Specification[T]] = None, limit: Option[Int] = None, offset: Option[Int] = None): Future[IndexedSeq[T]]
  def search[T <: DataSource : TypeTag](filter: Specification[T]): Future[IndexedSeq[T]] = {
    search(Some(filter), None, None)
  }

  def count[T <: DataSource : TypeTag](filter: Option[Specification[T]] = None): Future[Long]
  def count[T <: DataSource : TypeTag](filter: Specification[T]): Future[Long] = {
    count(Some(filter))
  }

  def exists[T <: DataSource : TypeTag](filter: Option[Specification[T]]): Future[Boolean]
  def exists[T <: DataSource : TypeTag](filter: Specification[T]): Future[Boolean] = {
    exists(Some(filter))
  }

  def create[T <: AggregateRoot : TypeTag](aggregates: Seq[T]): Future[Unit]

  def create[T <: AggregateRoot : TypeTag](aggregate: T): Future[Unit] = {
    create(Seq(aggregate))
  }

  def updatePairs[T <: AggregateRoot : TypeTag](pairs: Seq[(T, T)]): Future[Unit]

  def update[T <: AggregateRoot : TypeTag](oldAggregate: T, newAggregate: T): Future[Unit] = {
    updatePairs(Seq((oldAggregate, newAggregate)))
  }

  def update[T <: AggregateRoot : TypeTag](aggregate: T): Future[Unit] = {
    updatePairs(Seq((null.asInstanceOf[T], aggregate)))
  }

  def update[T <: AggregateRoot : TypeTag](aggregates: Seq[T]): Future[Unit] = {
    updatePairs(aggregates.map(it => (null.asInstanceOf[T], it)))
  }

  def delete[T <: AggregateRoot : TypeTag](aggregates: Seq[T]): Future[Unit]

  def delete[T <: AggregateRoot : TypeTag](aggregate: T): Future[Unit] = {
    delete(Seq(aggregate))
  }

  def submit[T <: DomainEvent : TypeTag](events: Seq[T]): Future[Unit]

  def submit[T <: DomainEvent : TypeTag](event: T): Future[Unit] = {
    submit(Seq(event))
  }

  def populate[T : TypeTag](report: Report[T]): Future[T]
}
